### ISSUE TITLE
lazy produces TypeScript error: `JSX element type 'Component' does not have any construct or call signatures`

### ISSUE THREAD

**Issue by birtles** [preactjs/preact#3219]
- [ ] Check if updating to the latest Preact version resolves the issue (it does not -- 10.5.13 is fine, but 10.5.14 is broken for me)

**Describe the bug**

As of 10.5.14 code like the following:

```ts
const MyComponent = lazy(() =>
  import('./my-component').then((m) => {
    return {
      default: m.MyComponent,
    };
  })
);
```

may begin to error with: `TS2604: JSX element type 'MyComponent' does not have any construct or call signatures`

#3139 changed the type of `lazy` as follows:

```diff
- export function lazy<T>(loader: () => Promise<{ default: T }>): T;
+ export function lazy<T>(loader: () => Promise<{ default: T } | T>): T;
```

From what I can tell, in the above code fragment, TypeScript tries to infer the type of `T` by matching the return value of the function against the newly added `T` instead of `{ default: T }` and hence resolves the return type of `lazy` as `{ default: React.FunctionComponent<Props> }` _not_ `React.FunctionComponent<Props>` as it previously did.

For the time being it seems possible to work around this by explicitly providing the generic type parameter such as follows:

```ts
import type { Props as MyComponentProps } from './my-component';

const MyComponent = lazy<React.FunctionComponent<MyComponentProps>>(() =>
  import('./my-component').then((m) => {
    return {
      default: m.MyComponent,
    };
  })
);
```

**Comment by JoviDeCroock** [comment]
This has been implemented in #3139

**Comment by birtles** [comment]
> This has been implemented in #3139

Is that right? #3139 is the regressing issue, not the fix.

**Comment by JoviDeCroock** [comment]
You are completely right
